text/0032-upgrade-bootstrapper.md

```diff
@@ -0,0 +1,153 @@
+
+Under normal circumstances, the following steps will be performed:
+
+1. the `upgrade-bootstrapper` is started
+2. if the path to the node's `config.toml` was provided via an arg, the `upgrade-bootstrapper` writes the path value to
+   a hard-coded, known location (if writing fails, it panics).  Otherwise it reads the path value from the known
+   location.  If the path cannot be read from the known location, it panics
+3. the `upgrade-bootstrapper` spawns a child process, running the `casper-node` binary in validator mode, passing it the
+   path to its `config.toml` as an arg
+4. while they are running, before the network reaches the upgrade point, the next version of the `casper-node` gets
+   installed, but not run.  The binary is always named `casper-node-next` and its config named `config-next.toml`.  Each
+   gets installed to the same location as the current equivalent files.  The location of the `config.toml` can be
+   retrieved from the known location used by the `upgrade-bootstrapper` at step 2.  An upgrade file for chainspec
+   modifications should also be installed and its path included in the new `config-next.toml`
+5. when the current upgrade activation point is reached (see [the first unresolved question](#unresolved-questions)
+   regarding how the running node will get to know of the activation point) and the node has finished all its work in
+   that era, it exits normally with an exit code 0
+6. the `upgrade-bootstrapper` renames
+     * `casper-node-next` to `casper-node` (replaces the old binary)
+     * `config.toml` to `config-old.toml`
+     * `config-next.toml` to `config.toml`
+7. the `upgrade-bootstrapper` runs the new `casper-node` in migrate-data mode (see below for further details), passing
+   it the path to `config.toml` and `config-old.toml` as args.  The node performs any migration of stored data and
+   config options required, then exits normally with an exit code 0
+8. repeat from step 3
+
+Should the child process stop with a non-zero exit code at any stage, the `upgrade-bootstrapper` should itself panic.
+It is up to the user to restart the `upgrade-bootstrapper` manually.
+
+It is assumed that the `casper-node` binary will live at a known location and that installing an upgraded binary will
+place the binary and its config file in a known location.  If we wish to be more flexible about this (particularly the
+location of the `casper-node` binary) we can handle these locations in a similar way to the location of the
+`casper-node`'s `config.toml`.
+
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### Changes to existing `casper-node` (running in validator mode)
+
+We require the following changes to `casper-node`:
+
+* running in validator mode, when we reach the upgrade activation point, exit gracefully
+* running in validator mode, if the node starts, joins and finds that it has already reached the upgrade activation
+  point, exit gracefully
```

Comment:
> Can a node/validator be late to obtain the updated code?

As long as we're relying on humans to actively download and install updated software, I'd say yes.  Probably even if we do all we can to automate this in the future, we should still plan to handle this case.

> If so, it is possible for them to 'miss' the upgrade activation point? What happens then?

Probably the consensus team will be better placed to answer whether we can handle that gracefully.  I was anticipating it would be like a validator joining/re-joining after the genesis timestamp has expired.

If they've missed multiple upgrades, ultimately the brute force approach of treating that case like a brand new user just starting for the first time with our updated software should suffice.

> Additionally, for new nodes that join the network after the first or second upgrade has been processed, what does the sync and joining process look like? Can we lay this out?

@xcthulhu and I spoke about this, and we believe that if we ensure the block header is such that it will always hold the protocol version and the list of validators for the given era (at least in the switch blocks) then we should be able to make use of the current design for fast sync.  It might be worth having a meeting with interested parties to ensure we're all on the same page though.

Reply:
It's not even required that the header holds the validator set as currently it doesn't and we're still able to sync – we read this information from the global state. Joining node will need a trusted hash that's after the last upgrade, that's all.